src/openfermion/resource_estimates/pbc/thc/utils/isdf.py

```diff
@@ -0,0 +1,1115 @@
+) -> npt.NDArray:
+    """Build k-point THC zeta (central tensor) for given Q, delta_G,
+    delta_G_prime.
+
+    Args:
+      mydf: instance of pyscf.pbc.df.FFTDF object.
+      Q: Momentum transfer (in 1BZ).
+      delta_G: Reciprocal lattice vector satisfying Q - (Q-k) = delta_G
+      delta_G_prime: Reciprocal lattice vector satisfying Q - (Q-k') = delta_G
+      grid_points: Real space grid. Dimension [num_grid_points, num_dim],
+        num_dim is 1, 2 or 3 for 1D, 2D, 3D.
+      xi_mu: array containing interpolating vectors determined during ISDF
+        procedure
+
+    Returns:
+      zeta: central tensor of dimension [num_interp_points, num_interp_points]
+
+    """
+    cell = df_inst.cell
+    num_grid_points = grid_points.shape[0]
+    # delta_G - delta_G_prime because we have Gpq and Gsr and Gsr = -Grs, phase
+    # = Delta G = Gpq + Grs
+    phase_factor = np.exp(-1j * (np.einsum(
+        "x,Rx->R", delta_G - delta_G_prime, grid_points, optimize=True)))
+    # Minus sign again due to we use Q = kp - kq, but we should have
+    # V(G + k_q - k_p)
+    coulG = tools.get_coulG(cell, k=-(Q + delta_G), mesh=df_inst.mesh)
+    weighted_coulG = coulG * cell.vol / num_grid_points
+    xi_muG = tools.fft(xi_mu.T, df_inst.mesh)
+    xi_muG *= weighted_coulG
+    vR = tools.ifft(xi_muG, df_inst.mesh)
+    zeta = np.einsum("R,Rn,mR->mn", phase_factor, xi_mu, vR, optimize=True)
+    return zeta
+
+
+def build_kpoint_zeta_single_tranlsation(
+        df_inst: df.FFTDF,
+        q: int,
+        delta_G: npt.NDArray,
+        grid_points: npt.NDArray,
+        xi_mu: npt.NDArray,
+) -> npt.NDArray:
+    """Build k-point THC zeta (central tensor) for given Q, delta_G,
+    delta_G_prime.
+
+    Args:
+      mydf: instance of pyscf.pbc.df.FFTDF object.
+      q: Momentum transfer kp-kq.
+      delta_G: Reciprocal lattice vector satisfying Q - (Q-k) = delta_G
+      grid_points: Real space grid. Dimension [num_grid_points, num_dim],
+        num_dim is 1, 2 or 3 for 1D, 2D, 3D.
+      xi_mu: array containing interpolating vectors determined during ISDF
+        procedure
+
+    Returns:
+      zeta: central tensor of dimension [num_interp_points, num_interp_points]
+
+    """
+    cell = df_inst.cell
+    num_grid_points = grid_points.shape[0]
+    # delta_G - delta_G_prime because we have Gpq and Gsr and Gsr = -Grs, phase
+    # = Delta G = Gpq + Grs
+    phase_factor = np.exp(
+        -1j * (np.einsum("x,Rx->R", delta_G, grid_points, optimize=True)))
+    # Minus sign again due to we use Q = kp - kq, but we should have
+    # V(G + k_q - k_p)
+    coulG = tools.get_coulG(cell, k=-q, mesh=df_inst.mesh)
+    weighted_coulG = coulG * cell.vol / num_grid_points
+    xi_muG = tools.fft(xi_mu.T, df_inst.mesh)
+    xi_muG *= weighted_coulG
+    vR = tools.ifft(xi_muG, df_inst.mesh)
+    zeta = np.einsum("R,Rn,mR->mn", phase_factor, xi_mu, vR, optimize=True)
+    return zeta
+
+
+def build_G_vectors(cell: gto.Cell) -> npt.NDArray:
```

Comment:
They are essentially doing the same thing, but in gvec logic everything is in terms of the underlying integers rather than the vectors here which are in absolute units (`1/L')